app/models/inventory_number.rb

```diff
@@ -1,16 +1,29 @@
 class InventoryNumber < ActiveRecord::Base
 
-  def self.all_codes
-    select("CAST(code AS integer) AS number").map(&:number)
+  validates :code, presence: true, uniqueness: true
+
+  def self.create_with_next_code!
+    self.create!(code: next_code)
+  end
+
+  def self.next_code
+    number = missing_code || (max_code + 1)
+    number.to_s.rjust(6, "0")
   end
 
-  def self.latest_code
-    select("CAST(code AS integer)").order("code desc").
-      first.try(:code).try(:to_i) || 1
+  def self.missing_code
+    sql_for_missing_code =
+      "SELECT s.i AS first_missing_code
+        FROM generate_series(1,#{max_code}) s(i)
```

Comment:
Suppose we have following codes in our `InventoryNumber` table: ["000001"," 000002", "999999"], as per this query we'll get `max_code` = 999999, then we generate series from 1 to 999999. and select least non existent number in `missing_code'.

How about taking `max_code` = `InventoryNumber.count` here? 

suppose we have 
["000001", ... , "000096", "000098", "000999", "999999"] 
As we can also enter InventoryNumber code manually , consider the last code entered was "999999" then total`InventoryNumber.count` = 10000 and we'll have to look for missing number between 1 to 100. 
It really not necessary to generate this series `generate_series(1,#{max_code})` from 1 to 999999. I think it is enough if we go from 1 to InventoryNumber.count. 

What do you think?, let me know your thoughts.